### PR TITLE
CMake: Unset cached results for C++11/14/17 detection if necessary

### PR DESCRIPTION
**Pull request by tamiko**
This pull request reorganizes some CMake stuff slightly and introduces a
"UNSET_IF_CHANGED" macro to ease some configuration logic. It looks a bit wild
but should be pretty straightforward.

Supersedes #6494
Reverts #6490

### CODE REVIEW

**Comment by tamiko** [comment]
@masterleinad @drwells Please have a look and check carefully!

File: cmake/checks/check_01_cxx_features.cmake [hunk 1]
```diff
@@ -35,6 +35,17 @@
...
 
+
+#
+#                              BIG DISCLAIMER
+#
+# Frankly speaking - this whole file is a mess and has to be rewritten and
+# restructured at some point. So, if you are the lucky one who wants to add
+# DEAL_II_WITH_CXX2X support, do not add yet another 300 lines of spaghetti
+# code but restructure the whole thing to something sane.
+#
```
Comment (drwells) [inline cmake/checks/check_01_cxx_features.cmake]: oh dear.

**Comment by tamiko** [comment]
@masterleinad But this is not necessary. The checks must produce the very same output.

**Comment by masterleinad** [comment]
The testsuite will not be helpful here. So let's merge.

**Comment by tamiko** [comment]
@masterleinad In case you try to change compiler versions the CMake will bark at you :-D

**Pull request merged**